### PR TITLE
[FLOC-1520] Create container

### PR DESCRIPTION
**Pull request by itamarst**
https://clusterhq.atlassian.net/browse/FLOC-1520

### CODE REVIEW

File: flocker/control/httpapi.py [hunk 1]
```diff
@@ -414,6 +408,48 @@ def state_datasets(self):
...
+        """
+        Create a new dataset in the cluster configuration.
+
+        :param unicode hostname: The address of the node on which the container
+            will run.
+
+        :param unicode name: A unique identifier for the container.
+
+        :param unicode image: The name of the Docker image to use for the
+            container.
```
Comment (wallrj) [inline flocker/control/httpapi.py]: Perhaps say "Name and optionally tag"

File: flocker/control/httpapi.py [hunk 2]
```diff
@@ -414,6 +408,48 @@ def state_datasets(self):
...
+
+        :return: A ``EndpointResponse`` describing the container which has
+            been added to the cluster configuration.
+        """
+        # Check if container by this name already exists, if it does
+        # return error.
```
Comment (wallrj) [inline flocker/control/httpapi.py]: We guess you're going to check configuration only here.

File: flocker/control/httpapi.py [hunk 3]
```diff
@@ -414,6 +408,48 @@ def state_datasets(self):
...
+
+        # Find node by hostname, creating new one if it doesn't exist
+        # (code can be shared with create_container_configuration).
```
Comment (wallrj) [inline flocker/control/httpapi.py]: Presumably `create_dataset_configuration`

**Comment by wallrj** [comment]
Looks good. 

Guess the tests will mirror those of the volume API endpoints.

And given that this can be hooked up to the zfs convergence agent, will there be an acceptance test?

Please proceed.

**Comment by itamarst** [comment]
An acceptance test would probably want to use "list current state", so we'd write it once that was done later.

File: flocker/control/httpapi.py [hunk 4]
```diff
@@ -39,6 +41,9 @@
         SCHEMA_BASE.child(b'endpoints.yml').getContent()),
     }
 
+CONTAINER_NAME_COLLISION = make_bad_request(
+    code=CONFLICT, description=u"The container name specified already exists."
```
Comment (jongiddy) [inline flocker/control/httpapi.py]: Does this really need the word "specified"? Can probably leave out.

File: flocker/control/httpapi.py [hunk 5]
```diff
@@ -414,6 +420,70 @@ def state_datasets(self):
...
+        schema_store=SCHEMAS
+    )
+    def create_container_configuration(self, host, name, image):
+        """
+        Create a new dataset in the cluster configuration.
+
+        :param unicode host: The address of the node on which the container
+            will run.
+
+        :param unicode name: A unique identifier for the container.
```
Comment (jongiddy) [inline flocker/control/httpapi.py]: Indicate that it is unique within the Flocker cluster, not just on the named host.

File: flocker/control/httpapi.py [hunk 6]
```diff
@@ -414,6 +420,70 @@ def state_datasets(self):
...
+
+        :param unicode image: The name of the Docker image to use for the
+            container.
+
+        :return: A ``EndpointResponse`` describing the container which has
+            been added to the cluster configuration.
```
Comment (jongiddy) [inline flocker/control/httpapi.py]: A -> An

File: flocker/control/test/test_httpapi.py [hunk 7]
```diff
@@ -173,6 +173,169 @@ def _build_app(test):
...
+                    Node(hostname=self.NODE_B),
+                }
+            )
+            self.assertEqual(deployment, expected)
+
+        saving.addCallback(created)
+        return saving
+
+    def test_configuration_updated_new_node(self):
+        d = self.assertResponseCode(
```
Comment (jongiddy) [inline flocker/control/test/test_httpapi.py]: docstring

File: flocker/control/test/test_httpapi.py [hunk 8]
```diff
@@ -173,6 +173,169 @@ def _build_app(test):
...
+                    ),
+                }
+            )
+            self.assertEqual(deployment, expected)
+
+        d.addCallback(created)
+        return d
+
+    def test_response(self):
+        container_json = {
```
Comment (jongiddy) [inline flocker/control/test/test_httpapi.py]: docstring

**Comment by jongiddy** [comment]
Looks good. Some minor inline comments and some failing spellchecks on Buildbot, but can be merged whern those are addressed.

**Pull request merged**